language/vm/src/check_bounds.rs

```diff
@@ -23,506 +23,477 @@ impl<'a> BoundsChecker<'a> {
+    fn check_struct_def(&self, struct_def: &StructDefinition, errors: &mut Vec<VMStatus>) {
+        check_bounds_impl(
+            &self.module.struct_handles,
+            struct_def.struct_handle,
+            errors,
+        );
+        // check signature (type) and type parameter for the field type
+        if let StructFieldInformation::Declared(fields) = &struct_def.field_information {
+            let type_param_count = match self
+                .module
+                .struct_handles
+                .get(struct_def.struct_handle.into_index())
+            {
+                Some(sh) => sh.type_parameters.len(),
+                None => 0,
+            };
+            // field signatures are inlined
+            for field in fields {
+                check_bounds_impl(&self.module.identifiers, field.name, errors);
+                self.check_type(&field.signature.0, errors);
+                self.check_type_parameter(&field.signature.0, type_param_count, errors);
+            }
+        }
     }
-}
 
-impl BoundsCheck<&CompiledModuleMut> for FunctionSignature {
-    #[inline]
-    fn check_bounds(&self, module: &CompiledModuleMut) -> Vec<VMStatus> {
-        self.return_types
-            .iter()
-            .map(|token| token.check_bounds((module, self)))
-            .chain(
-                self.arg_types
-                    .iter()
-                    .map(|token| token.check_bounds((module, self))),
-            )
-            .flatten()
-            .collect()
+    fn check_function_def(&self, function_def: &FunctionDefinition, errors: &mut Vec<VMStatus>) {
+        check_bounds_impl(&self.module.function_handles, function_def.function, errors);
+        for ty in &function_def.acquires_global_resources {
+            check_bounds_impl(&self.module.struct_defs, *ty, errors);
+        }
+        self.check_code(function_def, errors);
     }
-}
 
-impl LocalsSignature {
-    fn check_type_parameters(&self, type_formals_len: usize) -> Vec<VMStatus> {
-        self.0
-            .iter()
-            .flat_map(|ty| ty.check_type_parameters(type_formals_len))
-            .collect()
-    }
+    fn check_code(&self, function_def: &FunctionDefinition, errors: &mut Vec<VMStatus>) {
+        if function_def.is_native() {
+            return;
+        }
+        let code_unit = &function_def.code;
+        check_bounds_impl(&self.module.signatures, code_unit.locals, errors);
+
+        let type_param_count = match self
+            .module
+            .function_handles
+            .get(function_def.function.into_index())
+        {
+            Some(fh) => fh.type_parameters.len(),
+            None => 0,
+        };
+        // if there are locals check that the type parameters in local signature are in bounds.
+        let locals = match &self.module.signatures.get(code_unit.locals.into_index()) {
+            Some(locals) => &locals.0,
+            None => return, // stop now
```

Comment:
right and we don't want to crash because it's on entry and I changed and, in fact, crashed the test :) 
There is a conversation to have about what guarantees you can rely on but it may have a lot of implications...
I hope we can talk about what guarantees should the code have and when at some point soon, but most likely not too soon